transport/transport.go

```diff
@@ -0,0 +1,95 @@
+package transport
+
+import (
+	"context"
+	"net"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/mux"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+// DialTimeout is the maximum duration a Dial is allowed to take.
+// This includes the time between dialing the raw network connection,
+// protocol selection as well the handshake, if applicable.
+var DialTimeout = 60 * time.Second
+
+// AcceptTimeout is the maximum duration an Accept is allowed to take.
+// This includes the time between accepting the raw network connection,
+// protocol selection as well as the handshake, if applicable.
+var AcceptTimeout = 60 * time.Second
+
+// UpgradedConn is a connection that is is an extension of the net.Conn interface that provides multiaddr
+// information, and an accessor for the transport used to create the conn
+type UpgradedConn interface {
```

Comment:
This may not have gone through the "upgrader" and really shouldn't be called `UpgradedConn`. Any reason not to just call it a `TransportConn` or a `Conn`?

We'll have to rethink this anyways with the switch refactor but I don't want to imply any mandatory relationship with the upgrader.

Reply:
I found this confusing, fwiw. 

Before I really started really reading the go code, I was thinking of transport connections as more or less "raw" network connections and thought that the upgrader would upgrade *from* transport connections to secure + multiplexed connections. 

Looking at the implementations it seems that you pass an upgrader into the transport constructor, and it's the transport's job to apply it to the raw connections. That's fine, but the relationship between transports and the upgrader isn't clear from reading the Transport interface.

Reply:
In my head, I take the term "transport" mean things like TCP, UDP, WSS, WebRTC, QUIC, etc, so `TransportConn` wouldn't necessary denote that the connection has the capabilities required by libp2p, I'd take it to represent a raw tarnsport connection.

What I want to convey is that the libp2p stack _expects_ _**this**_ connection to support multiplexing, and to be encrypted, whether native or shimmed.

Reply:
I agree this is all really confusing. I just find calling this `UpgradedConn` even more confusing.

Really, It doesn't matter _that_ much as we'll need to revisit this in the switch refactor anyways.

Reply:
Just some thoughts. In C++ implementation we try to keep transport without any dependencies on Upgrader or UpgradedConnections. To this end we want to create a separate folder for ConnUpgrader and everything related to the upgrader (including UpgradedConn if we have one in future). You might consider this approach as well

Reply:
I'm also having second thoughts on calling this `UpgradedConn`.

@kamilsa – we also don't have a relationship between the connection and the upgrader, so this naming is definitely misleading if it made you think so. Thanks for the input. Note that the upgrader is a convenience component – it has no associated abstraction, but definitely plays a key role in shimming multiplexing and security in transports that do not support it natively.

Alternative names, off the top of my head:
* `CapableConn`: denotes that it has the capabilities libp2p requires.
* `P2PConn`

Reply:
More on the naming bikeshed:

* `Libp2pTransportConn`
* `P2PTransportConn`
* Just leave it as `transport.Conn`?

Regardless, `transport.Listener` should match (e.g, `UpgradedListener`).

Reply:
We had consensus around `CapableConn` -- I just renamed it. @raulk, @yusefnapora, @Stebalien signed off on it. @vyzo had some reservations, but I think it'll grow on people. If after a reasonable period people are uncomfortable, we can rename and alias it.